controllers/devworkspace/solver/che_routing.go

```diff
@@ -208,212 +270,297 @@ func isSecureScheme(scheme string) bool {
-			name := getEndpointExposingObjectName(machineName, workspaceID, port, endpointName)
-			var prefix string
-			var serviceURL string
-
-			prefix = getPublicURLPrefix(workspaceID, machineName, port, endpointName)
-			serviceURL = getServiceURL(port, workspaceID, namespace)
+func addEndpointToTraefikConfig(componentName string, e dw.Endpoint, cfg *traefikConfig, cheCluster *v2alpha1.CheCluster, routing *dwo.DevWorkspaceRouting) {
+	var routeName string
+	if e.Attributes.GetString(uniqueEndpointAttributeName, nil) == "true" {
+		// if endpoint is unique, we're exposing on /componentName/<endpoint-name>
+		routeName = e.Name
+	} else {
+		// if endpoint is NOT unique, we're exposing on /componentName/<port-number>
+		routeName = strconv.Itoa(e.TargetPort)
+	}
+	prefix := fmt.Sprintf("/%s/%s", componentName, routeName)
+	rulePrefix := fmt.Sprintf("PathPrefix(`%s`)", prefix)
 
-			rtrs[name] = traefikConfigRouter{
-				Rule:        fmt.Sprintf("PathPrefix(`%s`)", prefix),
-				Service:     name,
-				Middlewares: calculateMiddlewares(name),
-				Priority:    100,
-			}
+	// skip if exact same route is already exposed
+	for _, r := range cfg.HTTP.Routers {
+		if r.Rule == rulePrefix {
+			return
+		}
+	}
 
-			srvcs[name] = traefikConfigService{
-				LoadBalancer: traefikConfigLoadbalancer{
-					Servers: []traefikConfigLoadbalancerServer{
-						{
-							URL: serviceURL,
-						},
-					},
-				},
-			}
+	name := fmt.Sprintf("%s-%s-%s", routing.Spec.DevWorkspaceId, componentName, routeName)
+	cfg.HTTP.Routers[name] = traefikConfigRouter{
+		Rule:        rulePrefix,
+		Service:     name,
+		Middlewares: calculateMiddlewares(name, false),
+		Priority:    100,
+	}
 
-			mdls[name+"-prefix"] = traefikConfigMiddleware{
-				StripPrefix: &traefikConfigStripPrefix{
-					Prefixes: []string{prefix},
+	cfg.HTTP.Services[name] = traefikConfigService{
+		LoadBalancer: traefikConfigLoadbalancer{
+			Servers: []traefikConfigLoadbalancerServer{
+				{
+					URL: fmt.Sprintf("http://127.0.0.1:%d", e.TargetPort),
 				},
-			}
+			},
+		},
+	}
 
-			if infrastructure.IsOpenShift() {
-				mdls[name+"-auth"] = traefikConfigMiddleware{
-					ForwardAuth: &traefikConfigForwardAuth{
-						Address: "http://127.0.0.1:8089?namespace=" + namespace,
-					},
-				}
+	cfg.HTTP.Middlewares[name+"-prefix"] = traefikConfigMiddleware{
+		StripPrefix: &traefikConfigStripPrefix{
+			Prefixes: []string{prefix},
+		},
+	}
 
-				mdls[name+"-header"] = traefikConfigMiddleware{
-					Plugin: &traefikPlugin{
-						HeaderRewrite: &traefikPluginHeaderRewrite{
-							From:   "X-Forwarded-Access-Token",
-							To:     "Authorization",
-							Prefix: "Bearer ",
-						},
-					},
-				}
-			}
+	if infrastructure.IsOpenShift() {
+		cfg.HTTP.Middlewares[name+"-auth"] = traefikConfigMiddleware{
+			ForwardAuth: &traefikConfigForwardAuth{
+				Address: fmt.Sprintf("http://%s.%s:8089?namespace=%s", gateway.GatewayServiceName, cheCluster.Namespace, routing.Namespace),
```

Comment:
is https possible here? would prefer it otherwise potentially valid token can be sniffed from the unecrypted traffic.

Reply:
We're on http since beginning of the gateway https://github.com/eclipse-che/che-operator/blob/main/controllers/devworkspace/solver/che_routing.go#L506 and even with che-server workspaces https://github.com/eclipse-che/che-server/blob/main/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/external/TraefikGatewayRouteConfigGenerator.java#L60

It's communication inside of the cluster, so I would assume it's safe, but maybe I'm too naive. We can definitely do it, but it would need more work as we need all gateway components to listen on https. I would do it in separate issue.